### PR TITLE
ADDR-1930 Google PAD Integration Sample

### PR DESCRIPTION
**Pull request by kirstenpilla**
Shows how to load assets from the 3 different delivery types (install-time, fast-follow, on-demand).
In SampleScene there are 3 buttons for each delivery type. Clicking on the button will either load or unload a prefab. 

Code:
- Custom build script: BuildScriptPlayAssetDelivery.cs
- Custom AssetBundleProvider: PlayAssetDeliveryAssetBundleProvider.cs 
- Custom schema: PlayAssetDeliverySchema.cs
- Setup & Configures Addressables: AddressablesInitSingleton.cs
- Loads/unloads assets when button is pressed: LoadObject.cs
- Asset Pack Settings object: CustomAssetPackSettings.cs
- Asset Pack Settings object custom inspector: CustomAssetPackSettingsInspector.cs
- Serializes data to be stored in a json: CustomAssetPackData.cs
- Helper class: CustomAssetPackUtility.cs

README contains new instructions for this sample project.

How SampleScene looks like if all 3 assets are loaded:
image

How the custom Schema appears in the Inspector: 
image

"Manage Asset Packs" will open the Custom Asset Pack Settings: 
image

### CODE REVIEW

File: Advanced/Play Asset Delivery/Assets/PlayAssetDelivery/Runtime/InitializeAddressables.cs
Comment (juliusunity) [inline Advanced/Play Asset Delivery/Assets/PlayAssetDelivery/Runtime/InitializeAddressables.cs:92]: Users should never have to hardcode this pack name in their code. This is the first sign that something is being done in an unsupported way and might cause problems in the future.
I don't know exactly how addressables set up works, but just to make sure we understand each other - bundles created by addressables package will end up in streaming assets location (no matter where it is) only if they were dropped in project's "Assets/StreamingAssets" directory. If the bundle was added in a directory ending with .androidpack, then that bundle will end up in a separate asset pack, not in streaming assets path. Our VFS might be able to locate the asset in it's own install-time asset pack even when streaming assets path is provided, however it might not work in all cases and could stop working in the future without anyone in VFS / mobile team being aware that it would affect addressables.

File: Advanced/Play Asset Delivery/Assets/PlayAssetDelivery/Runtime/InitializeAddressables.cs
Comment (juliusunity) [inline Advanced/Play Asset Delivery/Assets/PlayAssetDelivery/Runtime/InitializeAddressables.cs:93]: The supported way of accessing streaming assets path is by using Application.streamingAssetsPath API ( https://docs.unity3d.com/Manual/StreamingAssets.html ) which will return the correct streaming assets path which can be fed to UnityWebRequest no matter where those streaming assets are actually placed (in base application, in data asset pack (either install-time or fast-follow delivered), in separate streaming assets asset pack (either install-time or fast-follow delivered).
I am not sure if this code is actually depending on the VFS behavior of fallbacking to correct "split" to load the asset bundle from the install-time delivered custom asset pack or if the code actually tries to get a path to the asset bundle in the streaming assets location, in which case Application.streamingAssetsPath is the correct path to use.

File: Advanced/Play Asset Delivery/Assets/PlayAssetDelivery/Runtime/InitializeAddressables.cs
Comment (juliusunity) [inline Advanced/Play Asset Delivery/Assets/PlayAssetDelivery/Runtime/InitializeAddressables.cs:101]: If I understand correctly, as I explained in the PlayAssetDeliveryAssetBundleProvider.cs script, this is depending on the incidental behavior which is not officially supported.

**Review by juliusunity — COMMENTED** [review]
Four comments for the whole PR:
* Very nice sample overall
* The sample seams to depend on the incidental behavior of our VFS when accessing bundles in custom install-time asset packs. This is problematic as it might stop working at any time.
* The sample packs 1 bundle per asset pack, however according to Play requirements, a single app can have only 50 asset packs at most. I don't know what is the philosophy of addressables regarding the amount of bundles per app, but I assume it is not limited, which might cause problems for the users. You should consider to add the functionality of having multiple bundles placed in a single asset pack.
* Scripts are in PlayAssetDelivery namespace. From what I know currently, it should not conflict with anything, however since this name is specific to Google's Play store, some of the Google's plugin might start using this namespace which could be problematic in the future. I would suggest to use Addressables.PlayAssetDelivery namespace or similar to avoid future problems.

File: Advanced/Play Asset Delivery/Assets/PlayAssetDelivery/Runtime/PlayAssetDeliveryAssetBundleProvider.cs
Comment (kirstenpilla) [reply Advanced/Play Asset Delivery/Assets/PlayAssetDelivery/Runtime/PlayAssetDeliveryAssetBundleProvider.cs:29]: Makes sense! I'll remove custom install-time packs from the sample

File: Advanced/Play Asset Delivery/Assets/PlayAssetDelivery/Runtime/InitializeAddressables.cs
Comment (kirstenpilla) [reply Advanced/Play Asset Delivery/Assets/PlayAssetDelivery/Runtime/InitializeAddressables.cs:92]: Hmm Addressables has a custom build processor (AddressablesPlayerBuildProcessor.cs) that temporarily moves the build folder (Library/com.unity.addressables) containing bundles, catalog.json, settings.json to Assets/StreamingAssets. Simply removing the bundles from the build folder (what I basically did in BuildScriptPlayAssetDelivery.cs) will prevent them from being added to the streaming assets path. 

By default Addressables expects to locate content (bundles, catalog.json, settings.json) in the Application.streamingAssetsPath directory. I'm guessing these files won't exist in Application.streamingAssetsPath if the streaming assets pack is fast-follow or on-demand. 

Is there any other way to retrieve the streaming assets pack name? AndroidAssetPacks.GetCoreUnityAssetPackNames() can return it, but not sure if there is a way to tell which pack name is the right one.

**Comment by kirstenpilla** [comment]
Thanks @juliusunity for the in-depth feedback! 

> * The sample packs 1 bundle per asset pack, however according to Play requirements, a single app can have only 50 asset packs at most. I don't know what is the philosophy of addressables regarding the amount of bundles per app, but I assume it is not limited, which might cause problems for the users. You should consider to add the functionality of having multiple bundles placed in a single asset pack.

Yeah you are right we don't enforce a bundle limit. I saw that the [Google sample project](https://developer.android.com/guide/playcore/asset-delivery/integrate-unity?language=api) recommended assigning 1 bundle to 1 asset pack, so I decided to follow that. I suppose we can map one asset pack to multiple bundles by writing to a file, and then at runtime read from it.

**Comment by juliusunity** [comment]
> Thanks @juliusunity for the in-depth feedback!
> 
> > * The sample packs 1 bundle per asset pack, however according to Play requirements, a single app can have only 50 asset packs at most. I don't know what is the philosophy of addressables regarding the amount of bundles per app, but I assume it is not limited, which might cause problems for the users. You should consider to add the functionality of having multiple bundles placed in a single asset pack.
> 
> Yeah you are right we don't enforce a bundle limit. I saw that the [Google sample project](https://developer.android.com/guide/playcore/asset-delivery/integrate-unity?language=api) recommended assigning 1 bundle to 1 asset pack, so I decided to follow that. I suppose we can map one asset pack to multiple bundles by writing to a file, and then at runtime read from it.

Yeah, I think Google went with the simplest use-case, however from what I've seen in the past, some projects have hundreds of small asset bundles, so for them having an ability to pack multiple bundles into one asset pack would be crucial.

**Comment by juliusunity** [comment]
I had some additional thought over night and I think the best (and simplest) solution for the install-time delivered content would be to simply not create any custom install-time asset packs with bundles created by addressables. Instead those bundles which a user would like to be delivered at install-time could be placed in "Assets/StreamingAssets" directly. This way your loading code will always know where to find bundles for which our AndroidAssetPacks API does not return a valid status / path. You would simply get Application.streamingAssetsPath and concatenate it with the bundle name. Passing that path to UnityWebRequest would work just fine.
This solution would be perfect because:
* Not dependant on unsupported behaviors
* Does not have potential performance issues
* Does not create additional asset packs (which is important due to 50 asset packs per application limit)
* From the users's perspective works exactly the same (no matter if one or multiple install-time) asset packs are created since they all would be installed all the time anyway, so no saving of space on the device by downloading only needed bundles.

One drawback (or rather a thing to be aware of) is that if there are many other assets in the project, streaming assets might get packed with a fast-follow delivery type, however I don't think that would be a problem for this type of use-case. So I suggest to consider simply dropping bundles into Assets/StreamingAssets directory when users choose to have them install-time delivered.

Oh one more thing. In my previous comments for the README file I wrote that "split application binary" is not necessary for this sample. If you will decide to use streaming assets for "install-time asset packs" logic, then that option will be needed (specifically when the user chooses to deliver any bundles during the install time.

**Comment by kirstenpilla** [comment]
> One drawback (or rather a thing to be aware of) is that if there are many other assets in the project, streaming assets might get packed with a fast-follow delivery type, however I don't think that would be a problem for this type of use-case. So I suggest to consider simply dropping bundles into Assets/StreamingAssets directory when users choose to have them install-time delivered.

Sounds good. Even if users do not choose to have install-time asset packs, Addressables will still have some files (catalog.json, settings.json) placed in Assets/StreamingAssets. So for our case we can never allow the streaming assets pack to be fast-follow delivery? Or in other words if the user exceeds the 1GB core Unity asset pack size threshold, they need to make sure that their streaming assets pack is always larger than the core data pack?

File: Advanced/Play Asset Delivery/Assets/PlayAssetDelivery/Runtime/AssetPackDownloadUtility.cs.meta
Comment (kirstenpilla) [reply Advanced/Play Asset Delivery/Assets/PlayAssetDelivery/Runtime/AssetPackDownloadUtility.cs.meta:1]: The script was deleted and I forgot to delete the meta file. Good catch

File: Advanced/Play Asset Delivery/Assets/PlayAssetDelivery/Runtime/LoadObject.cs
Comment (kirstenpilla) [reply Advanced/Play Asset Delivery/Assets/PlayAssetDelivery/Runtime/LoadObject.cs:34]: I'll rename the function to be less confusing. But currently "Load" is called when a button is pressed, which will either load or unload an object

**Comment by juliusunity** [comment]
> Sounds good. Even if users do not choose to have install-time asset packs, Addressables will still have some files (catalog.json, settings.json) placed in Assets/StreamingAssets. So for our case we can never allow the streaming assets pack to be fast-follow delivery? Or in other words if the user exceeds the 1GB core Unity asset pack size threshold, they need to make sure that their streaming assets pack is always larger than the core data pack?
Well, I don't know if recommending users to make sure that streaming assets pack is larger than the data pack is beneficial in any way. No matter which way the streaming assets pack is delivered, the Application.streamingAssetsPath will always return the correct path which you will be able to pass to UnityWebRequest, so there is no specific need for the customers to do project adaptations to ensure streaming assets pack ends up to be larger than data asset pack.
The drawback I had in mind was that in Addressables settings users would chose "install-time" delivery, however the pack could end up with a "fast-follow" delivery, therefore technically ignoring the user's setting. But in my opinion it is not bad (especially if it will be properly documented, so users would know what to expect). 
As for Addressables configuration files - adding json files to streaming assets is totally fine, you will just have to ensure that  AndroidAssetPacks.coreUnityAssetPacksDownloaded returns true before accessing those files and you will be always fine. I think the code in this project was doing exactly that - it was checking if that API returns true and in case it wasn't doing that, your code was downloading asset packs returned by AndroidAssetPacks.GetCoreUnityAssetPackNames. This is perfectly fine and will always work.

**Review by juliusunity — APPROVED** [review]
The sample looks great. I've left a few comments, but overall it is great. Well done.

File: Advanced/Play Asset Delivery/Assets/PlayAssetDelivery/Editor/BuildScriptPlayAssetDelivery.cs
Comment (kirstenpilla) [reply Advanced/Play Asset Delivery/Assets/PlayAssetDelivery/Editor/BuildScriptPlayAssetDelivery.cs:60]: Looks like a typo, it should be: AssetDatabase.CreateFolder(PackContentRootDirectory, m_AssetPackFolderName);

Comment (juliusunity) [reply Advanced/Play Asset Delivery/Assets/PlayAssetDelivery/Editor/BuildScriptPlayAssetDelivery.cs]: Hmm, are you sure it was a typo? PackContentRootDirectory is $"{m_RootDirectory}/{m_AssetPackFolderName}", so now this "CreateFolder" call will try to create $"{m_RootDirectory}/{m_AssetPackFolderName}/{m_AssetPackFolderName}" directory, which is probably not correct. Also the original concern was that looking at AssetDatabase.CreateFolder it seems that if the directory specified in the first parameter does not exist, the directory name specified in the second parameter is not created either. So after your change, even if duplicating m_AssetPackFolderName in the path is what we want, this folder will not be created, since at this point PackContentRootDirectory will always be empty (since it is deleted in a line above). This is the concern I had in my original comment.

Comment (kirstenpilla) [reply Advanced/Play Asset Delivery/Assets/PlayAssetDelivery/Editor/BuildScriptPlayAssetDelivery.cs]: Oh nevermind, yes you are right. The m_RootDirectory (Assets/PlayAssetDelivery) will always exist in the project. It also contains all the script files. Deleting and creating the PackContentRootDirectory(Assets/PlayAssetDelivery/CustomAssetPackContent) for each new build is the behavior we want. It will contain all the asset pack folders and bundles assigned to the custom asset packs.

**Review by ObsessiveNerd — APPROVED** [review]

**Review by paulb-unity — APPROVED** [review]
Great work!

**Comment by kirstenpilla** [comment]
Spoke with Andy and this PR (#946) modifies the behavior of AddressablesPlayerBuildProcessor which this sample project relies on to assign content to the streaming assets pack. I added a custom processor (PlayAssetDeliveryBuildProcessor) as a workaround, but we will need know which Addressables package version includes that PR. Will need to wait for the PR to land before this sample project can merged in unfortunately : (

**Comment by mrcarriere** [comment]
@kirstenpilla Is there an ETA on this? We're investigating our options right now, but it seems like this is going to block our distribution to Android.

**Comment by kirstenpilla** [comment]
> @kirstenpilla Is there an ETA on this? We're investigating our options right now, but it seems like this is going to block our distribution to Android.

Hi! The PR is likely to merge in around 2+ weeks since it is dependent on the aforementioned PR. In the mean time this PR is in a pretty good state as-is with some minor changes as needed. You can probably just pull down the branch and work off that.

**Comment by optimisez** [comment]
Hi @kirstenpilla. Any plan to integrate this Google PAD feature directly into Addressables package in future?

**Comment by Nardo23** [comment]
:0

**Pull request merged**
